dali/sasha/saxref.cpp

```diff
@@ -623,58 +632,70 @@ static void mergeDirPerPartDirs(cDirDesc *parent, cDirDesc *dir, const char *cur
+            continue;
+        }
+
+        std::string fname = fileItr->first;
 
-                cFileDesc *movedFile = nullptr;
-                for (unsigned k=0;k<file->N;k++)
+        cFileDesc *movedFile = nullptr;
+        unsigned numFileParts = file->N; // file may be deleted during for loop, save the number of parts
+        for (unsigned k=0;k<numFileParts;k++)
+        {
+            auto dirPerPartDirItr = parent->dirs.find(std::to_string(k+1)); // If not end(), dir-per-part directory k+1
+            if (dirPerPartDirItr == parent->dirs.end())
+                continue;
+
+            cDirDesc *dirPerPartDir = dirPerPartDirItr->second.get(); // dir-per-part directory k+1 under parent directory
+            auto dirPerPartFileItr = dirPerPartDir->files.find(fname);
+            if (dirPerPartFileItr != dirPerPartDir->files.end())
+            {
+                cFileDesc *dirPerPartFile = dirPerPartFileItr->second.get();
+                if (movedFile == nullptr)
                 {
-                    cDirDesc *dirPerPartDir = parent->dirs.find(std::to_string(k+1).c_str(), false);
-                    if (dirPerPartDir)
-                    {
-                        cFileDesc *dirPerPartFile = dirPerPartDir->files.find(fname,false);
-                        if (dirPerPartFile)
-                        {
-                            if (movedFile == nullptr)
-                            {
-                                parent->files.add(dirPerPartFile);
-                                movedFile = dirPerPartFile;
-                                movedFile->isDirPerPart = true;
-                            }
-                            else
-                            {
-                                movedFile->setpresent(0, k);
-                                if (dirPerPartFile->testmarked(0, k))
-                                    movedFile->setmarked(0, k);
-                            }
-                            dirPerPartDir->files.remove(dirPerPartFile);
-                        }
-                    }
+                    // Move the file from dirPerPartDir to parent
+                    movedFile = dirPerPartFileItr->second.get();
+                    parent->files[fname] = std::move(dirPerPartFileItr->second);
+                    movedFile->isDirPerPart = true;
+                }
+                else
+                {
+                    movedFile->setpresent(0, k);
+                    if (dirPerPartFile->testmarked(0, k))
+                        movedFile->setmarked(0, k);
                 }
+                // Delete cFileDesc from part directories since it has been moved to parent
+                // If deleting current file part, adjust iterator
+                if (dirPerPartFileItr == fileItr)
+                    fileItr = dirPerPartDir->files.erase(dirPerPartFileItr);
+                else
+                    dirPerPartDir->files.erase(dirPerPartFileItr);
             }
         }
```

Comment:
if the inner for loop does not find/move the file in 'dir' , then fileIter is never updated ?
I think you need to flag that it was updated (by erase) if it was,, and here at the end if not do:
```
if (!currentErased)
    fileIter++;